### PR TITLE
Add module documentation for `ops`

### PR DESCRIPTION
**Pull request by albi3ro**
Previously, most objects defined inside `qml.ops` were imported top-level, so there was less need for module-only documentation.  With the addition of developer-facing objects to the `ops` module, now we need to add a page for that module.

### CODE REVIEW

File: pennylane/devices/default_qubit.py [hunk 1]
```diff
@@ -490,13 +490,14 @@ def expval(self, observable, shot_range=None, bin_size=None):
         # to a component that can be re-used by devices as needed.
         if observable.name in ("Hamiltonian", "SparseHamiltonian"):
             assert self.shots is None, f"{observable.name} must be used with shots=None"
-
+            print("Im a hamiltonain")
```
Comment (Jaybsoni) [inline pennylane/devices/default_qubit.py:493]:
```suggestion
```

File: pennylane/devices/default_qubit.py [hunk 2]
```diff
@@ -490,13 +490,14 @@ def expval(self, observable, shot_range=None, bin_size=None):
...
             backprop_mode = (
                 not isinstance(self.state, np.ndarray)
                 or any(not isinstance(d, (float, np.ndarray)) for d in observable.data)
             ) and observable.name == "Hamiltonian"
 
             if backprop_mode:
+                print("I'm in backprop mode")
```
Comment (Jaybsoni) [inline pennylane/devices/default_qubit.py:500]:
```suggestion
```

File: pennylane/devices/default_qubit.py (hunk 1 above)
Comment (albi3ro) [reply pennylane/devices/default_qubit.py:493]: good catch 😆

**Review by Jaybsoni — COMMENTED** [review]
Looks good! I am just not sure about the naming. The name `qml.ops` suggests that this section of the documentation should cover all the classes mentioned in the `pennylane/ops/` folder. Maybe we can re-name that aspect or include the other headings like `qubit_ops`, `cv_ops`, etc.

**Comment by dime10** [comment]
Hmm, now I realize that this ops module introduces an inconsistency because `eigvals, generator, matrix, adjoint, ctrl` are all already listed under the `qml` module. It looks like the other API module sections only present those functions and classes which can only be accessed via the submodule name (e.g. `qml`**`.tape`**`.stop_recording`).
Is there any functionality that we invoke via `qml.ops.`?


> Looks good! I am just not sure about the naming. The name `qml.ops` suggests that this section of the documentation should cover all the classes mentioned in the `pennylane/ops/` folder. Maybe we can re-name that aspect or include the other headings like `qubit_ops`, `cv_ops`, etc.

Yes originally I thought it would be good to add the gates to this section, but I realized they already live in `qml.` (on top of being listed in the "Quantum operators" guide).

File: doc/code/qml_ops.rst [hunk 3]
```diff
@@ -0,0 +1,13 @@
+qml.ops
+=======
+
+Functions
+---------
```
Comment (dime10) [inline doc/code/qml_ops.rst:5]: This name is actually a bit confusing in this context because module members are often included under a `Functions` and a `Classes` header, but in this case the header refers to the module name.

**Comment by Jaybsoni** [comment]
Should be in good shape now! We should probably wait till [Prod](https://github.com/PennyLaneAI/pennylane/pull/2625) is merged so we can include it in the docs before merging this PR

File: pennylane/ops/op_math/__init__.py [hunk 4]
```diff
@@ -14,10 +14,36 @@
...
     :toctree: api
 
+    ~Adjoint
+    ~ControlledOperation
+    ~Controlled
+    ~ControlledOp
+    ~Pow
+    ~Sum
+    ~SProd
+    ~SymbolicOp
```
Comment (Jaybsoni) [inline pennylane/ops/op_math/__init__.py:45]:
```suggestion
    ~Adjoint
    ~ControlledOperation
    ~Controlled
    ~ControlledOp
    ~Pow
    ~Prod
    ~Sum
    ~SProd
    ~SymbolicOp
```

File: pennylane/ops/op_math/__init__.py [hunk 5]
```diff
@@ -14,10 +14,36 @@
...
+
+.. currentmodule:: pennylane.ops.op_math
+
+.. autosummary::
+    :toctree: api
+
+    ~adjoint
+    ~ctrl
+    ~op_sum
+    ~s_prod
```
Comment (Jaybsoni) [inline pennylane/ops/op_math/__init__.py:28]:
```suggestion
    ~adjoint
    ~ctrl
    ~op_sum
    ~prod
    ~s_prod
```

File: doc/introduction/operations.rst [hunk 6]
```diff
@@ -42,25 +42,45 @@ Operator functions
...
+1.2*(Hadamard(wires=[0]) + PauliX(wires=[0]) @ PauliZ(wires=[1]))
+
+Operator information functions
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+.. autosummary::
+
     ~pennylane.matrix
     ~pennylane.eigvals
     ~pennylane.generator
```
Comment (josh146) [inline doc/introduction/operations.rst:73]: Note that generator should return an operator as well

Comment (josh146) [reply doc/introduction/operations.rst:73]: I guess the main distinction is 'functions that return operators' vs. 'functions that don't return operators (could be arrays, strings, etc.)'

File: pennylane/ops/functions/__init__.py [hunk 7]
```diff
@@ -13,6 +13,16 @@
 # limitations under the License.
 """
 This module contains functions that act on operators and tapes.
```
Comment (josh146) [inline pennylane/ops/functions/__init__.py:15]: Do you think it might be confusing having separate `ops/functions` and `ops/op_math` going forward? There might be edge cases where it's not clear to a developer which file to add the feature to.

Comment (albi3ro) [reply pennylane/ops/functions/__init__.py:15]: It's already confusing.  The `simplify` function could go in either.

File: pennylane/ops/op_math/__init__.py [hunk 8]
```diff
@@ -14,10 +14,38 @@
...
+.. currentmodule:: pennylane.ops.op_math
+
+.. autosummary::
+    :toctree: api
+
+    ~adjoint
+    ~ctrl
+    ~op_sum
+    ~prod
+    ~s_prod
```
Comment (josh146) [inline pennylane/ops/op_math/__init__.py:29]: I assume all of these are top level imports (e.g., `qml.adjoint()`, `qml.ctrl()`, etc.)? If so, care needs to be taken here:

- The generated autosummary table should link to the top level `qml.adjoint()` docstrings
- We don't want Sphinx to generate _multiple_ docstrings for the same function, e.g., `qml.adjoint()` and `qml.ops.op_math.adjoint()`

File: pennylane/ops/op_math/controlled_class.py [hunk 9]
```diff
@@ -98,6 +112,20 @@ class Controlled(SymbolicOp):
 
     """
 
+    # pylint: disable=no-self-argument
+    @operation.classproperty
+    def __signature__(cls):  # pragma: no cover
+        # this method is defined so inspect.signature returns __init__ signature
+        # instead of __new__ signature
+        # See PEP 362
```
Comment (josh146) [inline pennylane/ops/op_math/controlled_class.py:120]: nice!

Comment (albi3ro) [reply pennylane/ops/op_math/controlled_class.py:120]: black magic required more black magic, but at least it should be working now.

File: pennylane/ops/op_math/__init__.py (hunk 8 above)
Comment (albi3ro) [reply pennylane/ops/op_math/__init__.py:29]: Yep. These should all be top-level imports. That was why I was getting errors about `s_prod` earlier in this PR.

File: doc/introduction/operations.rst (hunk 6 above)
Comment (Jaybsoni) [reply doc/introduction/operations.rst:73]: Would it be sufficient to just move generator up? Or do you think the title / description should change. Personally, the description might be a bit off as I don't really think of transforming an op to its generator as much as asking for the generator of the op

File: pennylane/ops/functions/__init__.py (hunk 7 above)
Comment (josh146) [reply pennylane/ops/functions/__init__.py:15]: Agree - should we just merge them? Have a single place for generic functions of operators?

File: pennylane/ops/op_math/__init__.py (hunk 8 above)
Comment (josh146) [reply pennylane/ops/op_math/__init__.py:29]: Got it 👍 Yeah noticed the latter happening:

https://pennylane--2730.org.readthedocs.build/en/2730/code/api/pennylane.prod.html

https://pennylane--2730.org.readthedocs.build/en/2730/code/api/pennylane.ops.op_math.prod.html

Comment (Jaybsoni) [reply pennylane/ops/op_math/__init__.py:29]: Ahh nice catch!

Comment (Jaybsoni) [reply pennylane/ops/op_math/__init__.py:29]: I replaced the links for to auto summary to specifically point to `~.pennylane.ops.op_math.prod`

File: pennylane/ops/functions/__init__.py (hunk 7 above)
Comment (albi3ro) [reply pennylane/ops/functions/__init__.py:15]: We tend to group things in PennyLane by what type of object they are: functions, classes, tranforms, etc.  We should instead start grouping things by dependencies, interactions, and responsibilities. But there's no good, clear cut answer on where to put things, especially in a language as flexible as Python.

Comment (josh146) [reply pennylane/ops/functions/__init__.py:15]: Yep for sure, it is not an easy problem to solve. My short term thinking was to avoid developer overhead that we already see with operations vs. templates ("should I add this to operations or templates?")

Comment (Jaybsoni) [reply pennylane/ops/functions/__init__.py:15]: Re-working the docs might be a good epic to add in the future since I imagine this will require moving multiple functions and classes around.

File: doc/introduction/operations.rst (hunk 6 above)
Comment (Jaybsoni) [reply doc/introduction/operations.rst:73]: Generator moved up 👍🏼

File: pennylane/ops/op_math/__init__.py (hunk 8 above)
Comment (Jaybsoni) [reply pennylane/ops/op_math/__init__.py:29]: Swapped back to top level, and routed the docstring pages in `ops/op_math/` to top level as well 👍🏼

**Comment by Jaybsoni** [comment]
@albi3ro, @AlbertMitjans, should we also include documentation on `simplify` in the `ops/op_math` page? I think its a good spot to mention it for now?

File: doc/introduction/operations.rst [hunk 10]
```diff
@@ -42,25 +42,45 @@ Operator functions
 ------------------
 
 Various functions and transforms are available for manipulating operators,
-and extracting information.
+and extracting information. These can be broken down into two main categories:
+
+Operator to Operator functions
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
Comment (Jaybsoni) [inline doc/introduction/operations.rst:48]:
```suggestion
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

File: doc/introduction/operations.rst [hunk 11]
```diff
@@ -42,25 +42,45 @@ Operator functions
...
+
+These operator functions act on operators to produce new operators.
+
+>>> op = qml.prod(qml.PauliX(0), qml.PauliZ(1))
+>>> op = qml.op_sum(qml.Hadamard(0), op)
+>>> op = qml.s_prod(1.2, op)
+1.2*(Hadamard(wires=[0]) + PauliX(wires=[0]) @ PauliZ(wires=[1]))
+
+Operator to Other functions
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
Comment (Jaybsoni) [inline doc/introduction/operations.rst:68]:
```suggestion
^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

File: doc/introduction/operations.rst [hunk 12]
```diff
@@ -42,25 +42,45 @@ Operator functions
...
+^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+.. autosummary::
+
     ~pennylane.matrix
     ~pennylane.eigvals
-    ~pennylane.generator
 
+These operator functions act on operators and return other data types.
 All operator functions can be used on instantiated operators,
```
Comment (antalszava) [inline doc/introduction/operations.rst:76]:
```suggestion
All operator functions can be used on instantiated operators.
```

File: pennylane/ops/op_math/controlled_class.py [hunk 13]
```diff
@@ -41,6 +43,17 @@ class Controlled(SymbolicOp):
             length as ``control_wires``. Defaults to ``True`` for all control wires.
         work_wires (Any): Any auxiliary wires that can be used in the decomposition
 
+    .. note::
+        This class, ``Controlled``, denotes a controlled version of any inidividual operation.
```
Comment (antalszava) [inline pennylane/ops/op_math/controlled_class.py:47]:
```suggestion
        This class, ``Controlled``, denotes a controlled version of any individual operation.
```

File: pennylane/ops/op_math/sum.py [hunk 14]
```diff
@@ -116,7 +120,27 @@ class Sum(Operator):
...
+            def circuit(weights):
+                qml.RX(weights[0], wires=0)
+                qml.RY(weights[1], wires=1)
+                qml.CNOT(wires=[0, 1])
+                qml.RX(weights[2], wires=1)
+                return qml.expval(sum_op)
+
+        >>> weights = qnp.array([0.1, 0.2, 0.3], requires_grad=True)
+        >>> qml.grad(circuit)(weights)
+        tensor([-0.09347337, -0.18884787, -0.28818254], requires_grad=True)"""
```
Comment (antalszava) [inline pennylane/ops/op_math/sum.py:143]:
```suggestion
        tensor([-0.09347337, -0.18884787, -0.28818254], requires_grad=True)
        """
```

File: pennylane/ops/op_math/control.py [hunk 15]
```diff
@@ -83,6 +83,13 @@ class ControlledOperation(Operation):
         control_wires: A wire or set of wires.
         control_values: An int or list of ints indicating the values each control wire should
             take.
+
+    .. note::
+        Currently, the :func:`~.ctrl` tranform uses this class ``ControlledOperation``.  This class
+        wraps an entire :class:`pennylane.tape.QuantumTape`, and it is rarely supported for native device
+        execution.  See :class:`pennylane.ops.op_math.Controlled` for a more versatile controlled operation
+        that wraps a single target ``Operator``.
```
Comment (antalszava) [inline pennylane/ops/op_math/control.py:91]: What's the plan for the evolution of `ControlledOperation` and `Controlled`? It seems confusing to have both of these two classes in the long run.

File: pennylane/ops/op_math/controlled_class.py [hunk 16]
```diff
@@ -99,6 +112,20 @@ class Controlled(SymbolicOp):
 
     """
 
+    # pylint: disable=no-self-argument
+    @operation.classproperty
+    def __signature__(cls):  # pragma: no cover
+        # this method is defined so inspect.signature returns __init__ signature
```
Comment (antalszava) [inline pennylane/ops/op_math/controlled_class.py:118]: Why was this required?

**Review by antalszava — APPROVED** [review]
Looks good :+1: Just minor questions.

File: pennylane/ops/op_math/control.py (hunk 15 above)
Comment (Jaybsoni) [reply pennylane/ops/op_math/control.py:91]: Once all of the pl controlled ops (CNOT, CZ, etc.) inherit from and are integrated with Controlled, we can then start to deprecate it !

**Review by rmoyard — APPROVED** [review]
It looks good to me 👍  same question than Antal about the signature though

File: pennylane/ops/op_math/control.py (hunk 15 above)
Comment (albi3ro) [reply pennylane/ops/op_math/control.py:91]: So `ControlledOperation` uses the `_controlled` property to decompose itself.  `Controlled` currently does not have any decompositions.  Once we add some basic `decomposition` logic to `Controlled`, I think we should prefer using it.  

Right now, `Controlled` is hidden away, but it is much more versatile, as it defines matrices, eigenvalues, diagonalizing gates, parameter shift rules, etc.

File: pennylane/ops/op_math/controlled_class.py (hunk 16 above)
Comment (albi3ro) [reply pennylane/ops/op_math/controlled_class.py:118]: `inspect.signature` first looks at `__signature__`, then it looks at the call signature for `__new__`, then it looks at the call signature for `__init__`.

Since we are using `__new__` to add in `Operation` functionality, it has an altered call signature.  We don't want to use that in the documentation. We want to use `__init__`'s call signature instead.
